docs: improve custom validator documentation for async hooks (#6228)
fastify/fastify: docs/Reference/Validation-and-Serialization.md

> it should not be used for initial validation. Accessing databases during
> validation may lead to Denial of Service attacks. Use
> [Fastify's hooks](./Hooks.md) like `preHandler` for `async` tasks after validation.

> >
> > ⚠ **Important**: When using custom validators with async `preValidation` hooks,
> > validators **must return** `{error}` objects instead of throwing errors.
> > Throwing errors from custom validators will cause unhandled promise rejections
> > that crash your application when combined with async hooks. See the
> > [custom validator examples](#using-other-validation-libraries) below for the correct pattern.

Reviewer: ```suggestion
>
> When using custom validators with async `preValidation` hooks,
> validators **must return** `{error}` objects instead of throwing errors.
> Throwing errors from custom validators will cause unhandled promise rejections
> that crash the application when combined with async hooks. See the
> [custom validator examples](#using-other-validation-libraries) below for the correct pattern.
```

This is already inside a warning block, so the "important" bit is not needed.
Likewise, use of `you` should be avoided in the formal documentation, see the [documentation style guide](https://fastify.dev/docs/latest/Guides/Style-Guide/#when-to-avoid-the-second-person-you-as-the-pronoun).

```js
const Joi = require('joi')

// ❌ WRONG: This will crash your app with async preValidation hooks
fastify.setValidatorCompiler(({ schema }) => {
  return (data) => {
    const { error, value } = schema.validate(data)
    if (error) {
      throw error // This causes unhandled promise rejections!
    }
    return { value }
  }
})

> // ✅ CORRECT: Return error objects instead of throwing

Reviewer 2: @Fdawgs 

Do we have docs conventions for this type of correct VS incorrect examples?

Reviewer: ```suggestion
```

@jean-michelet No, but I think the incorrect examples are redundant. If documentation listed all the ways not to do something they'd be millions of lines long.

Reviewer 2: Yes, I agree

fastify.setValidatorCompiler(({ schema }) => {
  return (data) => {
    try {
      const { error, value } = schema.validate(data)
      if (error) {

>         return { error } // Return the error, don't throw it

Reviewer: ```suggestion
        return { error } // Return the error, do not throw it
```

}, handler)
```

##### Custom Validator Best Practices

When implementing custom validators, follow these patterns to ensure compatibility with all Fastify features:

> **✅ Always return objects, never throw:**
> ```js
> // CORRECT
> return { value: validatedData }  // On success
> return { error: validationError } // On failure
> ```
> 
> **❌ Never throw errors directly:**
> ```js
> // WRONG - will crash with async hooks
> throw new Error('Validation failed')
> ```

Reviewer: ```suggestion
** Always return objects, never throw:**
```js
return { value: validatedData }  // On success
return { error: validationError } // On failure
```
```

> **✅ Use try-catch for safety:**
> ```js
> fastify.setValidatorCompiler(({ schema }) => {
>   return (data) => {
>     try {
>       // Your validation logic here

Reviewer: ```suggestion
** Use try-catch for safety:**
```js
fastify.setValidatorCompiler(({ schema }) => {
  return (data) => {
    try {
      // Validation logic here
```

      const result = schema.validate(data)
      if (result.error) {
        return { error: result.error }
      }
      return { value: result.value }
    } catch (e) {
      // Catch any unexpected errors
      return { error: e }
    }
  }
})
```

> This pattern ensures your validators work correctly with both sync and async `preValidation` hooks, preventing unhandled promise rejections that can crash your application.

Reviewer: ```suggestion
This pattern ensures validators work correctly with both sync and async `preValidation` hooks, preventing unhandled promise rejections that can crash an application.
```
